### PR TITLE
[ULMS-1843] Send event notifications

### PR DESCRIPTION
**Pull request by mgrachev**

### CODE REVIEW

**Review by khodzha — APPROVED** [review]

File: src/app/ws/handler.rs [hunk 1]
```diff
@@ -122,6 +123,32 @@ async fn handle_socket<S: State>(socket: WebSocket, authn: Arc<ConfigMap>, state
     loop {
         tokio::select! {
             Ok(msg) = nats_rx.recv() => {
+                if let Some(headers) = msg.headers {
+                    let label = headers
```
Comment (manifest) [inline src/app/ws/handler.rs:127]: Consider replacing `PRESENCE_EVENT_LABEL`

with 
```
PRESENCE_MESSAGING_PATTERN: "unicast"
PRESENCE_RECIPIENT_AGENT_ID: "me.john-doe.example.org"
PRESENCE_SENDER_AGENT_ID: "me.john-doe.example.org"
```
to send a message (such as `label=agent.replaced`) to the particular agent.


and
```
PRESENCE_MESSAGING_PATTERN: "broadcast"
PRESENCE_SENDER_AGENT_ID: "me.john-doe.example.org"
```
to send a message (such as `label=agent.enter|agent.leave`) to all the agents within the room, but the sender.

Comment (manifest) [reply src/app/ws/handler.rs:127]: To state a clear intent and to generalize the message processing.

File: src/app/ws/handler.rs [hunk 2]
```diff
@@ -355,7 +411,7 @@ fn get_agent_id_from_token(
     Ok(agent_id)
 }
 
-fn make_response(response: Response) -> String {
+fn serialize_to_json<T: Serialize>(response: T) -> String {
```
Comment (manifest) [inline src/app/ws/handler.rs:414]: Why do we pass `T` by value here?

Comment (mgrachev) [reply src/app/ws/handler.rs]: Fixed.

File: src/app/ws/handler.rs (hunk 1 above)
Comment (mgrachev) [reply src/app/ws/handler.rs]: Fixed.

File: src/app/ws/event.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    AgentLeave,
+    #[serde(rename = "agent.replaced")]
+    AgentReplaced,
+    #[serde(rename = "agent.auth_timed_out")]
+    AuthTimedOut,
+    #[serde(rename = "agent.pong_timed_out")]
+    PongTimedOut,
+}
+
+impl ToString for Label {
```
[the remark this answers is not in the document]
Comment (mgrachev) [reply src/app/ws/event.rs]: Fixed.

File: src/test_helpers/state.rs [hunk 4]
```diff
@@ -1,34 +1,45 @@
...
+use uuid::Uuid;
 
 #[derive(Clone)]
 pub struct TestState {
     config: Config,
     db_pool: TestDb,
     authz: Authz,
-    replica_id: String,
+    replica_id: Uuid,
+    nats_client: Arc<dyn NatsClient>,
```
Comment (manifest) [inline src/test_helpers/state.rs:27]: It should probably be `Box` here.

File: src/session/map.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use tokio::sync::oneshot;
+
+pub type SessionValue = (SessionId, oneshot::Sender<()>);
+
+#[derive(Clone, Debug)]
+pub struct SessionMap {
+    inner: Arc<Mutex<HashMap<SessionKey, SessionValue>>>,
```
Comment (manifest) [inline src/session/map.rs:12]: Why did you decide to guard the `SessionMap` by `Mutex`?

File: src/db/agent_session.rs [hunk 6]
```diff
@@ -215,107 +206,96 @@ impl<'a> AgentCounter<'a> {
...
+    pub async fn execute(&self, conn: &mut PgConnection) -> sqlx::Result<AgentSessionId> {
+        sqlx::query_as!(
+            AgentSessionId,
             r#"
             UPDATE agent_session
-            SET outdated = $1
-            WHERE
-                agent_id = $2
-                AND classroom_id = $3
+            SET replica_id = $3
```
Comment (manifest) [inline src/db/agent_session.rs:264]: When do we need to update `replica_id` ?

File: src/test_helpers/state.rs (hunk 4 above)
Comment (mgrachev) [reply src/test_helpers/state.rs:27]: No, it must be `Arc` due the following error:
```
error[E0277]: the trait bound `dyn NatsClient: Clone` is not satisfied
  --> src/test_helpers/state.rs:27:5
   |
21 | #[derive(Clone)]
   |          ----- in this derive macro expansion
...
27 |     nats_client: Box<dyn NatsClient>,
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `dyn NatsClient`
```

File: chart/values.yaml [hunk 7]
```diff
@@ -70,6 +70,7 @@ service:
   ports:
     http: 8080
     metrics: 8081
+    internal: 8082
```
Comment (manifest) [inline chart/values.yaml:73]: We've discussed on `8081` for internal-http, `8888` for metrics.

File: src/app/session_manager.rs [hunk 8]
```diff
@@ -1,130 +1,92 @@
...
+    NotFound,
 }
 
 /// Manages agent sessions by handling incoming commands.
 /// Also, closes old agent sessions.
-pub fn run<S: State>(
-    state: S,
-    mut cmd_rx: mpsc::UnboundedReceiver<Command>,
+pub fn run(
+    // sessions: SessionMap,
```
Comment (manifest) [inline src/app/session_manager.rs:42]: A forgotten comment.

**Pull request merged**